tests/lib/rules/no-extra-parens.js

```diff
@@ -1403,6 +1403,25 @@ ruleTester.run("no-extra-parens", rule, {
         invalid("const [a = (b)] = []", "const [a = b] = []", "Identifier"),
         invalid("const {a = (b)} = {}", "const {a = b} = {}", "Identifier"),
 
+        // LHS of assigments/Assignment targets
+        invalid("(a) = b", "a = b", "Identifier"),
+        invalid("(a.b) = c", "a.b = c", "MemberExpression"),
+        invalid("(a) += b", "a += b", "Identifier"),
+        invalid("(a.b) >>= c", "a.b >>= c", "MemberExpression"),
+        invalid("[(a) = b] = []", "[a = b] = []", "Identifier"),
+        invalid("[(a.b) = c] = []", "[a.b = c] = []", "MemberExpression"),
+        invalid("({ a: (b) = c } = {})", "({ a: b = c } = {})", "Identifier"),
+        invalid("({ a: (b.c) = d } = {})", "({ a: b.c = d } = {})", "MemberExpression"),
+        invalid("[(a)] = []", "[a] = []", "Identifier"),
+        invalid("[(a.b)] = []", "[a.b] = []", "MemberExpression"),
+        invalid("[,(a),,] = []", "[,a,,] = []", "Identifier"),
+        invalid("[...(a)] = []", "[...a] = []", "Identifier"),
+        invalid("[...(a.b)] = []", "[...a.b] = []", "MemberExpression"),
+        invalid("({ a: (b) } = {})", "({ a: b } = {})", "Identifier"),
+        invalid("({ a: (b.c) } = {})", "({ a: b.c } = {})", "MemberExpression"),
+
+        // TODO: add tests for the RestElement in object patterns when it becomes supported in espree
```

Comment:
Isn't this syntax supported by Espree already? Or am I misunderstanding what this means?

```js
const { ...foo } = bar;
```

Reply:
> Isn't this syntax supported by Espree already? Or am I misunderstanding what this means?
> 
> ```js
> const { ...foo } = bar;
> ```

That works, but Espree doesn't allow parens around the target in rest properties. The missing test cases should be something like this:

```js
({ ...(a) } = bar);
({ ...(a.b) } = bar);
```

These parens seem to be a valid syntax in assignments (not in declarations), but that isn't supported by Espree.

So, the part of this PR that removes those parens is untestable with the default parser. That's the new `RestElement(node)` handler. It should work (most likely well but untested) with `babel-eslint` and `@typescript-eslint/parser` which both allow parens there.

The same code in  `RestElement(node)`  also removes parens around rest elements in array patterns:

```js
[...(a)] = []
```

That works in Espree and has test cases.

Maybe we should add some code to restrict `RestElement(node)` for now to work with rest elements in array patterns only?

Reply:
Understood, thanks! As long as the rule doesn't crash, I think it makes sense to consider this an upstream bug in Espree/Acorn and fix it there.

Reply:
Just tested with `@typescript-eslint/parser` and it seems to be working fine. The rule removes extra parens around `a` and `a.b`:

```js
/* eslint no-extra-parens: ["error"]*/

({ ...(a) } = {});
({ ...(a.b) } = {});
```

It turned out that nothing happens with `babel-eslint`. It's still `ExperimentalRestProperty` there. The rule doesn't report extra parens but also doesn't crash. These nodes are just ignored.

Also added a commit with TODO tests to make sure we don't forget what's this about.

It seems that this issue has been already reported at acornjs/acorn#872